### PR TITLE
Suitable for Ascend. Modify SSD and RetinaHead Model.

### PR DESCRIPTION
**Pull request by akstt**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Suitable for Ascend. Modify SSD and RetinaHead Model.

## Modification

- The main modification points are as follows:
  1. Create a new file `mmdet/models/dense_heads/ascend_anchor_head.py`. Create class AscendAnchorHead that inherits class AnchorHead.
  2. Create a new file `mmdet/models/dense_heads/ascend_retina_head.py`. Create class AscendRetinaHead that inherits class RetinaHead and class AscendAnchorHead. 
  3. Create a new file `mmdet/models/dense_heads/ascend_ssd_head.py`. Create class AscendSSDHead that inherits class SSDHead and class AscendAnchorHead. 
  4. Create a new file `mmdet/core/bbox/assigners/ascend_assign_result.py`. Create class AscendAssignResult that similar class AssignResult. 
  5. Create a new file `mmdet/core/bbox/assigners/ascend_max_iou_assigner.py`. Create class AscendMaxIoUAssigner that similar class MaxIoUAssigner. 
  6.  Create a new file `mmdet/utils/ascend_util.py`. Build function that suitable ascend.
  7.  Modify `mmdet/models/dense_heads/__init__.py`, `mmdet/core/bbox/assigners/__init__.py`, `mmdet/utils/__init__.py`. For new classes and functions can be able to be called into.

## BC-breaking (Optional)

None

## Use cases (Optional)

- test case
  - `tests/test_models/test_dense_heads/test_ascend_head.py`

## Checklist

- [x]  Pre-commit or other linting tools are used to fix the potential lint issues.
- [x]  The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+                 ),
+                 reg_decoded_bbox=False,
+                 train_cfg=None,
+                 test_cfg=None,
+                 init_cfg=dict(
+                     type='Xavier',
+                     layer='Conv2d',
+                     distribution='uniform',
+                     bias=0)):
+        super(AscendSSDHead, self).__init__(
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/ascend_ssd_head.py:81]: `super().__init__(xx=xx, xx=xx, xx=xx ...)`

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                    valid_flag_list,
+                    gt_bboxes_list,
+                    img_metas,
+                    gt_bboxes_ignore_list=None,
+                    gt_labels_list=None,
+                    label_channels=1,
+                    unmap_outputs=True,
+                    return_sampling_results=False,
+                    return_level=True):
+        return AscendAnchorHead.get_targets(
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/ascend_ssd_head.py:115]: does this mean `super().get_targets(xxx)`?

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        )
+
+    def concat_loss(self,
+                    concat_cls_score, concat_bbox_pred,
+                    concat_anchor, concat_labels,
+                    concat_label_weights,
+                    concat_bbox_targets, concat_bbox_weights,
+                    concat_pos_mask, concat_neg_mask,
+                    num_total_samples):
+        num_images, num_anchors, _ = concat_anchor.size()
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/ascend_ssd_head.py:136]: doc strings

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        concat_topk_loss_cls_neg, _ = torch.topk(concat_loss_cls_all * concat_neg_mask, k=num_anchors, dim=1)
+        concat_loss_cls_pos = torch.sum(concat_loss_cls_all * concat_pos_mask, dim=1)
+
+        anchor_index = torch.arange(end=num_anchors, dtype=torch.float, device=concat_anchor.device).view((1, -1))
+        topk_loss_neg_mask = (anchor_index < concat_num_neg_samples.view(-1, 1)).float()
+
+        concat_loss_cls_neg = torch.sum(concat_topk_loss_cls_neg * topk_loss_neg_mask, dim=1)
+        loss_cls = (concat_loss_cls_pos + concat_loss_cls_neg) / num_total_samples
+
+        if self.reg_decoded_bbox:
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/ascend_ssd_head.py:157]: if self.reg_decoded_bbox should be False, how about add a assert in init?

File: mmdet/core/bbox/assigners/ascend_assign_result.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+from mmdet.utils import util_mixins
+
+
+class AscendAssignResult(util_mixins.NiceRepr):
+    """Stores assignments between predicted and truth boxes.
+
+    Attributes:
```
Comment (BIGWangYuDong) [inline mmdet/core/bbox/assigners/ascend_assign_result.py:10]: args does not match parameters.

File: mmdet/models/dense_heads/ascend_anchor_head.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                            label_channels=1,
+                            unmap_outputs=True):
+        num_imgs, num_anchors, _ = concat_anchors.size()
+        # assign gt and sample concat_anchors
+        assign_result = self.assigner.assign(
+            concat_anchors, concat_gt_bboxes, concat_gt_bboxes_ignore,
+            None if self.sampling else concat_gt_labels,
+            concat_bboxes_ignore_mask=concat_valid_flags)
+
+        sampling_result = None
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/ascend_anchor_head.py:117]: this hardcode is a little bit strange

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            raise RuntimeError
+
+        loss_bbox_all = smooth_l1_loss(
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/ascend_ssd_head.py:160]: can we use self.bbox_loss to calculate the loss? Otherwise the argument/config will have no effect.

File: mmdet/models/dense_heads/ascend_ssd_head.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    def concat_loss(self,
+                    concat_cls_score, concat_bbox_pred,
+                    concat_anchor, concat_labels,
+                    concat_label_weights,
+                    concat_bbox_targets, concat_bbox_weights,
+                    concat_pos_mask, concat_neg_mask,
+                    num_total_samples):
+        num_images, num_anchors, _ = concat_anchor.size()
+
+        concat_loss_cls_all = F.cross_entropy(
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/ascend_ssd_head.py:138]: can we use self.cls_loss to calculate the loss? Otherwise, the argument/config will have no effect.

File: mmdet/models/dense_heads/ascend_ssd_head.py (hunk 7 above)
Comment (akstt) [reply mmdet/models/dense_heads/ascend_ssd_head.py:234]: Keep consistent with the original code. The argument/config have no effect in ssd loss function.

File: mmdet/models/dense_heads/ascend_ssd_head.py (hunk 8 above)
Comment (akstt) [reply mmdet/models/dense_heads/ascend_ssd_head.py:211]: Keep consistent with the original code. The argument/config have no effect in ssd loss function.

File: mmdet/models/dense_heads/ascend_ssd_head.py (hunk 2 above)
Comment (akstt) [reply mmdet/models/dense_heads/ascend_ssd_head.py:162]: AscendSSDHead inherit from SSDHead and AscendAnchorHead. super().get_targets(xxx) will call get_targets from SSDHead. I want call get_targets from  AscendAnchorHead

File: mmdet/models/dense_heads/ascend_ssd_head.py (hunk 4 above)
Comment (akstt) [reply mmdet/models/dense_heads/ascend_ssd_head.py:232]: Solved
